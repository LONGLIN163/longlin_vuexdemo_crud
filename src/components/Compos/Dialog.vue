<template>
    <el-dialog title="product info" :visible="dialogFormVisible" @close='toggleDialogForm'>

        <el-form :model="form">

            <h1>{{formMode}}</h1>

            <el-form-item v-if="formMode==='edit'" label="id" :label-width="formLabelWidth">
                <el-input v-model="form.id" autocomplete="off" disabled></el-input>
            </el-form-item>

            <el-form-item label="name" :label-width="formLabelWidth">
                <el-input v-model="form.name" autocomplete="off"></el-input>
            </el-form-item>

            <el-form-item label="description" :label-width="formLabelWidth">
                <el-input v-model="form.description" autocomplete="off"></el-input>
            </el-form-item>


            <el-form-item label="taste" :label-width="formLabelWidth">
                <el-select v-model="form.taste" placeholder="please select" style="width: 100%;">
                    <el-option
                    v-for="item in types"
                    :key="item"
                    :label="item"
                    :value="item">
                    </el-option>
                </el-select>

            </el-form-item>


            <el-form-item label="color" :label-width="formLabelWidth">
                <div class="colorBox">
                    <el-color-picker v-model="form.color"></el-color-picker>
                    <div class="colorTxt">{{form.color}}</div>
                </div>
            </el-form-item>

            <el-form-item label="image" :label-width="formLabelWidth">
                <el-input v-model="form.image" autocomplete="off"></el-input>
            </el-form-item>


            <el-form-item label="expires" :label-width="formLabelWidth">
                
                    <el-date-picker
                        v-model="form.expires"
                        type="date"
                        style="width: 100%;"
                        placeholder="choose a date"
                    >
                    </el-date-picker>
                
            </el-form-item>

            <el-form-item label="price" :label-width="formLabelWidth">
                <el-input v-model="form.price" autocomplete="off"></el-input>
            </el-form-item>

        </el-form>

        <div slot="footer" class="dialog-footer">
            <el-button @click="toggleDialogForm">Cancel</el-button>
            <el-button v-show="formMode===''" type="primary" @click="handleCreate(form)">Create</el-button>
        </div>

    </el-dialog>
</template>

<script>
    import store  from '@/vuex/store'
    import {mapState,mapMutations,mapActions} from 'vuex'

    export default {
        data(){
            return {
                formLabelWidth: '120px',
            }
        },
        props:['formMode','form'],
        computed:{
          ...mapState(['articleList','dialogFormVisible','types'])
        },
        methods: {
          handleCreate(form){
            this.createArticleAction(form);
            this.toggleDialogForm()
          },
          ...mapMutations(['toggleDialogForm']),
          ...mapActions(['createArticleAction'])
        },
        store
    }
</script>

<style scoped>
    .colorTxt{
        height: 40px;
        line-height: 40px;
        margin-left: 10px;
    }
    .colorBox{
        display: flex;
        flex-direction: row;
        align-items: flex-start
    }

</style>