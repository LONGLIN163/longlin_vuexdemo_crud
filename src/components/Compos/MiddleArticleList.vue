<template>
  <el-container>
    <el-main>
      <el-table
        :data="filteredList.filter(data => !search || data.name.toLowerCase().includes(search.toLowerCase()))"
        :default-sort = "{prop: 'id', order: 'descending'}"
        style="width: 100%"
      >
        <el-table-column type="expand">
          <template slot-scope="props">
            <el-row :gutter="20">
              <el-col :span="12">
                <div class="grid-content bg-purple">
                  <el-form label-position="left" inline class="demo-table-expand-left" >
                    <el-form-item label="ID:">
                      <span>{{ props.row.id }}</span>
                    </el-form-item>
                    <el-form-item label="Name:">
                      <span>{{ props.row.name }}</span>
                    </el-form-item>
                    <el-form-item label="Taste:">
                      <span>{{ props.row.taste }}</span>
                    </el-form-item>
                    <el-form-item label="Color:">
                      <span>{{ props.row.color }}</span>
                    </el-form-item>
                    <el-form-item label="Expires date:">
                      <span>{{ props.row.expires }}</span>
                    </el-form-item>
                    <el-form-item label="Price:">
                      <span>{{ props.row.price }}</span>
                    </el-form-item>

                  <el-button 
                      size="mini" 
                      type="danger" 
                      data-test="delArticleOnTable"
                      @click="delArticleAction(props.row.id)"
                    >Delete</el-button>

                  </el-form>
                </div>
              </el-col>
              <el-col :span="12">
                <div class="grid-content bg-purple">
                  <el-form label-position="left" inline class="demo-table-expand-right" >
                    <el-form-item label="Image:">
                      <span>
                        <el-image 
                          style="width: 400px;"
                          :src="props.row.image"
                        ></el-image>
                      </span>
                    </el-form-item>
                  </el-form>
                </div>
              </el-col>
            </el-row>
          </template>
        </el-table-column>

        <el-table-column
          label="Article ID"
          sortable
          prop="id">
        </el-table-column>
        <el-table-column
          label="Name"
          title='name'
          id='name'
          sortable
          prop="name">
        </el-table-column>
        <el-table-column
          label="Taste"
          sortable
          prop="taste">
        </el-table-column>
        <el-table-column
          label="Expires date"
          sortable
          prop="expires">
        </el-table-column>
        <el-table-column
          label="Price"
          sortable
          prop="price">
        </el-table-column>

        <el-table-column align="right">
          <template slot="header" slot-scope="scope">
            <el-input
              v-model="search"
              size="mini"
              placeholder="search..."/>
          </template>

          <template slot-scope="scope">
            <el-button
              size="mini"
              type="danger"
              data-test="delArticle"
              @click="delArticleAction(scope.row.id)"
              >Delete</el-button>
          </template>
        </el-table-column>
      </el-table>

      <Dialog 
        :formMode='formMode'
        :form='form'
        v-if="dialogFormVisible"
      ></Dialog>
      
      <el-header style="text-align: center; font-size: 12px; margin-top: 16px;">
        <el-button id="createBtn" type="primary" round @click="TOOGLE_DIALOGFORM">Create Article</el-button>
      </el-header>
      
    </el-main>
  </el-container>
</template>

<script>

    import Dialog  from '@/components/Compos/Dialog.vue'
    import {mapState,mapMutations, mapActions} from 'vuex'

    export default {
        data() {
          return {
            formMode:'', 
            form: {
                name: '',
                description: '',
                taste: '',
                color: '',
                image: '',
                isFruit: true,
                expires: '',
                price: ''
            },
            search: ''
          }
        },
        components:{
          Dialog
        },
        computed:{
          ...mapState(['dialogFormVisible','filteredList'])
        },
        methods: {
          ...mapMutations(['SHOW_SELECTED_ARTICLE','TOOGLE_DIALOGFORM']),
          ...mapActions(['delArticleAction'])
        }
    }
</script>

<style lang="scss" scoped>

  .demo-table-expand-left {
    padding-left: 50px;
  }

  .demo-table-expand-right {
    padding-left: 50px;
    border-left: 1px dashed $color-font-defalt;
  }
  .demo-table-expand-left label {
    width: 90px;
    color: $color-font-defalt;
    font-weight: bold;
  }
  .demo-table-expand-left .el-form-item {
    margin-right: 0;
    margin-bottom: 0;
    width: 100%;
  }

  #createBtn{
    background-color: $color-primary;
    border: none;
  }

  .descending{
    border-top-color: $color-hint !important;
  }
    


</style>