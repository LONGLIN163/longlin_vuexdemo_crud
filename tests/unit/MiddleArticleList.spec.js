import { mount,createLocalVue} from '@vue/test-utils'
import MiddleArticleList from '@/components/Compos/MiddleArticleList'
import Dialog from '@/components/Compos/Dialog'

import Vuex from 'vuex'
import ElementUI from "element-ui"
import {mockListData} from '../mockedData'

const localVue = createLocalVue();
localVue.use(ElementUI);  // install ElementUI to the local vue instance
localVue.use(Vuex); // install vuex to the local vue instance

let wrapper;
describe('Component MiddleArticleList tests:', () => {

  // define them outside of breforeEach
  let actions
  let state
  let store

  beforeEach(() => {

    // mock actions
    actions = {
      delArticleAction: jest.fn()
    }

    //mock state
    state={
      filteredList:mockListData,
      dialogFormVisible:false
    }

    // create a fake store
    store = new Vuex.Store({
      state,
      actions
    })

    
    wrapper = mount(MiddleArticleList,{
      localVue,
      store
    })
  })
  
  
  it('Get MiddleArticleList snapshot', async () => {
    expect(wrapper).toMatchSnapshot()
  })
    
  it('test if delArticleAction has been called when click del btn on table',async () => {
    const selectedItem=await wrapper.findAll('[data-test="delArticle"]').at(1);
    await selectedItem.trigger('click')
    expect(actions.delArticleAction).toHaveBeenCalled()
  })

  it('test if delArticleAction has been called when click del btn inside folded row',async () => {
    // click span anchor first
    const oneSpanAnchor=await wrapper.find('.el-table__body-wrapper tbody tr:first-child td:first-child');
    await oneSpanAnchor.trigger('click')
    // click del btn
    const selectedInnerItem=await wrapper.findAll('[data-test="delArticle"]').at(0);
    await selectedInnerItem.trigger('click')
    expect(actions.delArticleAction).toHaveBeenCalled()
  })

  it('test if component Dialog can be rendered',async () => {
    store.state.dialogFormVisible=true // can not use await here

    //Here need to create a new wrapper, because we can not test the compo controled by v-if
    const wrapper2 = mount(MiddleArticleList,{
      localVue,
      store
    })
    expect(wrapper2.findComponent(Dialog).exists()).toBe(true)
  })

  afterEach(() => {
    wrapper.destroy();
  });


})