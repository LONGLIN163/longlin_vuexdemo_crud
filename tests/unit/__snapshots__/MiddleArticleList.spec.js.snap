// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component MiddleArticleList tests: Get MiddleArticleList snapshot 1`] = `
<section class="el-container">
  <main class="el-main">
    <div class="el-table el-table--fit el-table--scrollable-x el-table--enable-row-hover el-table--enable-row-transition" style="width: 100%;">
      <div class="hidden-columns">
        <div></div>
        <div></div>
        <div title="name" id="name"></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
      </div>
      <div class="el-table__header-wrapper">
        <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 528px;">
          <colgroup>
            <col name="el-table_1_column_1" width="48">
            <col name="el-table_1_column_2" width="80">
            <col name="el-table_1_column_3" width="80">
            <col name="el-table_1_column_4" width="80">
            <col name="el-table_1_column_5" width="80">
            <col name="el-table_1_column_6" width="80">
            <col name="el-table_1_column_7" width="80">
          </colgroup>
          <thead class="">
            <tr class="">
              <th colspan="1" rowspan="1" class="el-table_1_column_1   undefined el-table__expand-column  is-leaf el-table__cell">
                <div class="cell"></div>
              </th>
              <th colspan="1" rowspan="1" class="el-table_1_column_2 descending    is-leaf is-sortable el-table__cell">
                <div class="cell">Article ID<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
              </th>
              <th colspan="1" rowspan="1" class="el-table_1_column_3     is-leaf is-sortable el-table__cell">
                <div class="cell">Name<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
              </th>
              <th colspan="1" rowspan="1" class="el-table_1_column_4     is-leaf is-sortable el-table__cell">
                <div class="cell">Taste<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
              </th>
              <th colspan="1" rowspan="1" class="el-table_1_column_5     is-leaf is-sortable el-table__cell">
                <div class="cell">Expires date<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
              </th>
              <th colspan="1" rowspan="1" class="el-table_1_column_6     is-leaf is-sortable el-table__cell">
                <div class="cell">Price<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
              </th>
              <th colspan="1" rowspan="1" class="el-table_1_column_7  is-right   is-leaf el-table__cell">
                <div class="cell">
                  <div class="el-input el-input--mini">
                    <!----><input type="text" autocomplete="off" placeholder="search..." class="el-input__inner">
                    <!---->
                    <!---->
                    <!---->
                    <!---->
                  </div>
                </div>
              </th>
            </tr>
          </thead>
        </table>
      </div>
      <div class="el-table__body-wrapper is-scrolling-left">
        <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 528px;">
          <colgroup>
            <col name="el-table_1_column_1" width="48">
            <col name="el-table_1_column_2" width="80">
            <col name="el-table_1_column_3" width="80">
            <col name="el-table_1_column_4" width="80">
            <col name="el-table_1_column_5" width="80">
            <col name="el-table_1_column_6" width="80">
            <col name="el-table_1_column_7" width="80">
          </colgroup>
          <tbody>
            <tr class="el-table__row">
              <td rowspan="1" colspan="1" class="el-table_1_column_1  undefined el-table__expand-column el-table__cell">
                <div class="cell">
                  <div class="el-table__expand-icon"><i class="el-icon el-icon-arrow-right"></i></div>
                </div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_1_column_2   el-table__cell">
                <div class="cell">5</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_1_column_3   el-table__cell">
                <div class="cell">apple</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_1_column_4   el-table__cell">
                <div class="cell">Sleek</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_1_column_5   el-table__cell">
                <div class="cell">2021-06-09T19:05:55.250Z</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_1_column_6   el-table__cell">
                <div class="cell">419.00</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_1_column_7 is-right  el-table__cell">
                <div class="cell"><button type="button" class="el-button el-button--danger el-button--mini" data-test="delArticle">
                    <!---->
                    <!----><span>Delete</span>
                  </button></div>
              </td>
            </tr>
            <tr class="el-table__row">
              <td rowspan="1" colspan="1" class="el-table_1_column_1  undefined el-table__expand-column el-table__cell">
                <div class="cell">
                  <div class="el-table__expand-icon"><i class="el-icon el-icon-arrow-right"></i></div>
                </div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_1_column_2   el-table__cell">
                <div class="cell">1</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_1_column_3   el-table__cell">
                <div class="cell">pineapple</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_1_column_4   el-table__cell">
                <div class="cell">Awesome</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_1_column_5   el-table__cell">
                <div class="cell">2022-10-13T19:10:51.621Z</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_1_column_6   el-table__cell">
                <div class="cell">346.00</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_1_column_7 is-right  el-table__cell">
                <div class="cell"><button type="button" class="el-button el-button--danger el-button--mini" data-test="delArticle">
                    <!---->
                    <!----><span>Delete</span>
                  </button></div>
              </td>
            </tr>
            <!---->
          </tbody>
        </table>
        <!---->
        <!---->
      </div>
      <!---->
      <!---->
      <!---->
      <!---->
      <div class="el-table__column-resize-proxy" style="display: none;"></div>
    </div>
    <!---->
    <header class="el-header" style="height: 60px; text-align: center; font-size: 12px; margin-top: 16px;"><button type="button" class="el-button el-button--primary is-round" id="createBtn">
        <!---->
        <!----><span>Create Article</span>
      </button></header>
  </main>
</section>
`;
